src/openslide-vendor-dicom.c

```diff
@@ -681,84 +660,6 @@ static bool add_level(openslide_t *osr,
-                "Missing DimensionOrganizationType");
-    return false;
-  }
-  if (g_str_equal(type, "TILED_FULL")) {
-    // used by eg. 3dhistech ... frames are in row-major order and we
-    // can simply use the BOT
-  } else if (g_str_equal(type, "3D") || g_str_equal(type, "TILED_SPARSE")) {
-    // used by eg. leica ... we need to read the position of every tile
-    l->frame_position = g_new(int64_t, l->tiles_across * l->tiles_down);
-    for (int64_t frame_number = 0;
-         frame_number < l->tiles_across * l->tiles_down;
-         frame_number++) {
-      l->frame_position[frame_number] = -1;
-    }
-
-    DcmSequence *per_frame_functional_groups_seq;
-    if (!get_tag_seq(f->metadata, PerFrameFunctionalGroupsSequence,
-                     &per_frame_functional_groups_seq)) {
-      g_set_error(err, OPENSLIDE_ERROR, OPENSLIDE_ERROR_FAILED,
-                  "Missing PerFrameFunctionalGroupsSequence");
-      return false;
-    }
-
-    uint32_t n_items = dcm_sequence_count(per_frame_functional_groups_seq);
-    for (uint32_t i = 0; i < n_items; i++) {
-      DcmError *dcm_error = NULL;
-      DcmDataSet *per_frame_functional_group =
-        dcm_sequence_get(&dcm_error, per_frame_functional_groups_seq, i);
-      if (!per_frame_functional_group) {
-        _openslide_dicom_propagate_error(err, dcm_error);
-        return false;
-      }
-
-      DcmDataSet *dimension_index_vals;
-      if (!get_tag_seq_item(per_frame_functional_group,
-                            FrameContentSequence, 0,
-                            &dimension_index_vals)) {
-        g_set_error(err, OPENSLIDE_ERROR, OPENSLIDE_ERROR_FAILED,
-                    "Missing FrameContentSequence");
-        return false;
-      }
-
-      uint32_t tag = dcm_dict_tag_from_keyword(DimensionIndexValues);
-      int64_t col, row;
-      DcmElement *values =
-        dcm_dataset_get(&dcm_error, dimension_index_vals, tag);
-      if (!values ||
-          !dcm_element_get_value_integer(&dcm_error, values, 0, &col) ||
-          !dcm_element_get_value_integer(&dcm_error, values, 1, &row)) {
-        _openslide_dicom_propagate_error(err, dcm_error);
-        return false;
-      }
-
-      // DICOM uses 1-based indexing, we want 0 based
-      // frame number i is the tile at (col, row)
-      if (col < 1 || row < 1) {
-        g_set_error(err, OPENSLIDE_ERROR, OPENSLIDE_ERROR_FAILED,
-                    "Invalid frame position (%"PRId64", %"PRId64")", col, row);
-        return false;
-      }
-      int64_t frame_index = (col - 1) + (row - 1) * l->tiles_across;
-      if (l->frame_position[frame_index] != -1) {
-        g_set_error(err, OPENSLIDE_ERROR, OPENSLIDE_ERROR_FAILED,
-                    "Duplicate frame (%"PRId64", %"PRId64")", col, row);
-        return false;
-      }
-      l->frame_position[frame_index] = i + 1;
-    }
-  } else {
-    g_set_error(err, OPENSLIDE_ERROR, OPENSLIDE_ERROR_FAILED,
-                "Unsupported DimensionOrganisationType");
```

Comment:
Is libdicom now doing this check internally?

Reply:
OK, I put this check back.

Reply:
Looks like it's gone again, I guess because of [this code](https://github.com/jcupitt/libdicom/commit/4828495d3c5c325fb258b9df915f7948e8d4fbde#diff-e4206ff375a29a4fc3867522531849694c83dddf198db2b3d602b3b0261d6decR709-R713)?  I agree that we should be trusting libdicom to know what it can and cannot read, rather than imposing our own judgement in OpenSlide.  But my reading of the libdicom code is that if the `DimensionOrganizationType` is something unexpected like `3D_TEMPORAL`, we'll try to compute the tile index as though it were `TILED_FULL`, and return random tiles.  Could we have one or the other library fail if the organization type is unrecognized?